### ISSUE TITLE
False-negative `type-abstract` in dictionary initialization

### ISSUE THREAD

**Issue by bersbersbers** [python/mypy#15264]
**Bug Report**

`type-abstract` is falsely missing when initializing a dictionary.

**To Reproduce**

```python
"""False-negative `type-abstract`."""
from abc import abstractmethod

class Class:
    @abstractmethod
    def method(self) -> None:
        pass

my_list: list[type[Class]] = []
my_list.append(Class)  # type-abstract

my_list_init: list[type[Class]] = [Class]  # type-abstract

my_dict: dict[int, type[Class]] = {}
my_dict[0] = Class  # type-abstract

my_dict_init: dict[int, type[Class]] = {0: Class}  # *no* type-abstract
```
https://mypy-play.net/?mypy=latest&python=3.11&gist=2f58d30cce7e6ddadfc2b0f28ef68395

**Expected Behavior**

main.py:10: error: Only concrete class can be given where "Type[Class]" is expected  [type-abstract]
main.py:12: error: Only concrete class can be given where "Type[Class]" is expected  [type-abstract]
main.py:15: error: Only concrete class can be given where "Type[Class]" is expected  [type-abstract]
**main.py:17: error: Only concrete class can be given where "Type[Class]" is expected  [type-abstract]**
Found 4 errors in 1 file (checked 1 source file)

**Actual Behavior**

main.py:10: error: Only concrete class can be given where "Type[Class]" is expected  [type-abstract]
main.py:12: error: Only concrete class can be given where "Type[Class]" is expected  [type-abstract]
main.py:15: error: Only concrete class can be given where "Type[Class]" is expected  [type-abstract]
Found 3 errors in 1 file (checked 1 source file)

**Your Environment**

- Mypy version used: 1.3.0
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.11.3

**Comment by sobolevn** [comment]
This happens because `check_arg` receives `tuple[Literal[0], () -> Class]` as a type argument in case of `{0: Class}`, while all other usages generate flat types.
